src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        provideCapacity(size + 1);
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkAddID(index);
+        provideCapacity(size + 1);
+        System.arraycopy(elements, index, elements,
+                index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkID(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkID(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkID(index);
+        T detail = elements[index];
+        System.arraycopy(elements, index
+                + 1, elements, index, --size - index);
+        return detail;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = indexOf(element);
+        if (index != ELEMENT_NOT_FOUND) {
+            return remove(index);
+        }
+        throw new NoSuchElementException("Detail not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkID(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index + "is out of bounds for size: " + size);
+        }
+    }
+
+    private void checkAddID(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index + "is out of bounds for size: " + size);
+        }
+    }
+
+    private void provideCapacity(int minCapacity) {
+        int currentCapacity = elements.length;
+        if (minCapacity > currentCapacity) {
+            int newCapacity = currentCapacity + (currentCapacity >> 1);
+            T[] newElements = (T[]) new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, size);
+            elements = newElements;
+        }
+    }
+
+    private int indexOf(T element) {
+        for (int i = 0; i < elements.length; i++) {
+            if (element == null) {
+                if (elements[i] == null) {
+                    return i;
+                }
+            } else if (element.equals(elements[i])) {
+                return i;
+            }
```

Comment:
Anastasia's comment not fixed
```suggestion
            if (element == elements[i] || element != null && element.equals(elements[i])) {
                 return i;
            }
```